snp_measurement/src/stage0.rs

```diff
@@ -0,0 +1,39 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::Context;
+use log::{debug, info};
+use sha2::{Digest, Sha256};
+use std::path::PathBuf;
+use x86_64::PhysAddr;
+
+/// The address of the first byte after the end of the firmware image.
+///
+/// The firmware image gets loaded just below the 4GiB boundary.
+pub const FIRMWARE_TOP: PhysAddr = PhysAddr::new(0x1_0000_0000);
```

Comment:
The BIOS gets shadowed in the low memory as well. I don't know exactly how QEMU handles this, but some other VMMs just copy the first 128K of memory to 1M-128K.

Reply:
Done.